### ISSUE TITLE
unused_parens false positive on braced macro calls

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust#78747]
I tried this code:

```rust
macro_rules! x {
    {} => { "q" };
}

fn f() -> String {
    (x! {}.to_string())
}

fn main() {}
```

I expected to see this happen: No warning about unnecessary parens

Instead, this happened: A warning about unnecessary parens suggesting to remove parens. Attempting to actually follow the advice causes an error.

```
warning: unnecessary parentheses around block return value
 --> src/main.rs:6:5
  |
6 |     (x! {}.to_string())
  |     ^^^^^^^^^^^^^^^^^^^ help: remove these parentheses
  |
  = note: `#[warn(unused_parens)]` on by default
```

This can be worked around by using `x!().to_string` or `(x! {}).to_string()` instead, however this shouldn't be necessary as the original code is fine.

This happens since Rust 1.42.

### Meta

`rustc --version --verbose`:
```
rustc 1.47.0 (18bf6b4f0 2020-10-07)
binary: rustc
commit-hash: 18bf6b4f01a6feaf7259ba7cdae58031af1b7b39
commit-date: 2020-10-07
host: x86_64-unknown-linux-gnu
release: 1.47.0
LLVM version: 11.0
```

**Comment by apiraino** [comment]
Assigning `P-medium` as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2378747.20unused_parens.20false.20positive.20on.20braced.20.E2.80.A6) and removing `I-prioritize`.

**Comment by Aaron1011** [comment]
Note that the parenthesis are required due to the fact that we have special handling for macro invocations with curly braces:

https://github.com/rust-lang/rust/blob/b2d115f6db5172c961dfeb50de15f35784dbc7c9/compiler/rustc_parse/src/parser/stmt.rs#L108-L111

Without the parenthesis, we parse `x!{}` as a statement, not an expression. However, `x!().to_string()` (without any surrounding parenthesis) compiles.

**Comment by kawadakk** [comment]
The suggestion is correct as of #88690 (included in 1.57.0).
